File: backend/src/models/user.model.js
import userSchema from "../lib/schema/user-schema.js";
import logger from "../helpers/logger.js";

// MASUKAN USER KE DATABASE
async function create({ id, username, email, password }) {
  try {
    return await userSchema.create({ id, username, email, password, photo: null });
  } catch (err) {
    logger.err(`UserModel -> createUser: ${err.message}`);
    throw err;
  }
}

// MENGAMBIL SATU DATA USER SESUAI EMAIL
async function findByEmail(email) {
  try {
    return await userSchema.findOne({ email });
  } catch (err) {
    logger.err(`UserModel -> findByEmail: ${err.message}`);
    throw err;
  }
}

// MENGAMBIL SATU DATA USER SESUAI DENGAN EMAIL DAN PASSWORD
async function findByEmailAndPassword(email, password) {
  try {
    return await userSchema.findOne({ email, password });
  } catch (err) {
    logger.err(`UserModel -> findByEmailAndPassword: ${err.message}`);
    throw err;
  }
}

// MENGAMBIL SEMUA USER SESUAI DENGAN EMAIL ATAU PASSWORD
async function findByEmailOrUsername(keyword) {
  try {
    return await userSchema
      .find({
        $or: [
          { email: { $regex: keyword, $options: "i" } },
          { username: { $regex: keyword, $options: "i" } },
        ],
      })
      .select("-password");
  } catch (err) {
    logger.err(`UserModel -> findByEmailOrUsername: ${err.message}`);
    throw err;
  }
}

async function deleteAllCollection(params) {
  try {
    
  } catch (err) {
    logger.err(`UserModel -> deleteAllCollection: ${err.message}`);
    throw err;
  }
}

export default {
  create,
  findByEmail,
  findByEmailAndPassword,
  findByEmailOrUsername,
};

File: frontend/lib/data/models/user_model.dart
class UserModel {
  final String id;
  final String email;
  final String password;
  final String username;
  final String? photo;

  UserModel({
    required this.id,
    required this.email,
    required this.password,
    required this.username,
    this.photo,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'] as String,
      email: json['email'] as String,
      password: json['password'] as String,
      username: json['username'] as String,
      photo: json['photo'] != null ? json['photo'] as String : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'email': email,
      'password': password,
      'username': username,
      'photo': photo,
    };
  }
}
